### ISSUE TITLE
Latest nightly issue

### ISSUE THREAD

**Issue by amircodota** [rust-lang/rust-clippy#12068]
### Summary

```
warning: unnecessary use of `to_string`
   --> language-servers/ls_installer/src/java/tests.rs:165:9
    |
165 |         locations[0].uri.to_string().split('?').next().unwrap(),
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `locations[0].uri.split('?')`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
    = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
```

but `uri` does not have a `split` method

### Lint Name

clippy::unnecessary_to_owned

### Version

```text
cargo +nightly --version                                                                                                                          
cargo 1.77.0-nightly (ac6bbb332 2023-12-26)
```

**Comment by samueltardieu** [comment]
Hi @amircodota and thanks for your report.

Could you please provide a minimal compilable example that reproduces the issue?

**Comment by amircodota** [comment]
Sure.

In my Cargo.toml I add a dependency on url

```
[dependencies]
url = "2.5.0"
```

This is my main.rs:

```rust
use url::Url;

fn main() {
    let url = Url::parse("file://file").unwrap();
    println!("{}", url.to_string().split('?').next().unwrap());
}
```

If I run `cargo clippy` all is good.
If I run `cargo +nightly clippy` I get

```
warning: unnecessary use of `to_string`
 --> src/main.rs:5:20
  |
5 |     println!("{}", url.to_string().split('?').next().unwrap());
  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `url.split('?')`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_to_owned
  = note: `#[warn(clippy::unnecessary_to_owned)]` on by default
```

AFAICT, there's no `split` method on `url::Url`

**Comment by samueltardieu** [comment]
Here is a self-contained example which does not require an external crate:

```rust
fn main() {
    struct S;
    impl AsRef<str> for S {
        fn as_ref(&self) -> &str {
            unimplemented!()
        }
    }
    impl ToString for S {
        fn to_string(&self) -> String {
            unimplemented!()
        }
    }
    let _ = S.to_string().split('?').next().unwrap();
}
```
```
13 |     let _ = S.to_string().split('?').next().unwrap();
   |             ^^^^^^^^^^^^^^^^^^^^^^^^ help: use: `S.split('?')`
```
